pkg/ebpf/bytecode/runtime/printk_patcher.go

```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+// PatchPrintkNewline patches log_debug calls to always print one newline, no matter what the kernel does.
+//
+// For context, in kernel 5.9.0, bpf_trace_printk adds a newline automatically to anything it prints
+// This means that when we support both older and newer kernels, bpf_printk is going to have
+// an inconsistent behavior. To avoid this, we have a wrapper called log_debug (see bpf_helpers_custom.h)
+// that adds a newline to the message before calling bpf_trace_printk. In older kernels
+// this ensures that a newline is added. In newer ones it would mean that two newlines are
+// added, so this patcher removes that newline in those cases.
+func PatchPrintkNewline(m *manager.Manager) error {
+	kernelVersion, err := kernel.HostVersion()
+	if err != nil {
+		return err // can't detect kernel version, don't patch
+	}
+	if kernelVersion < kernel.VersionCode(5, 9, 0) {
+		return nil // Do nothing in older kernels
+	}
+
+	progs, err := m.GetProgramSpecs()
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+
+	for _, p := range progs {
+		_, err := PatchPrintkInstructions(p)
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+// PatchPrintkInstructions patches the instructions of a program to remove the newline character
+// It's separated from PatchPrintkNewline so it can be tested independently, also so that we can
+// check how many patches are performed
+func PatchPrintkInstructions(p *ebpf.ProgramSpec) (int, error) {
+	var errs []error // list of errors that happened while patching, if any
+	numPatches := 0  // number of patches performed
+
+	// Compute some opcodes we'll need
+	movImmOpCode := asm.Mov.Op(asm.ImmSource)
+	ldDWImmOpCode := asm.LoadImmOp(asm.DWord)
+	movRegOpCode := asm.Mov.Op(asm.RegSource)
+	addImmOpCode := asm.Add.Op(asm.ImmSource)
+
+	// In some cases the compiler might reuse the same instruction for multiple calls to bpf_trace_printk,
+	// keep track of that to avoid errors when we don't find a newline in an instruction we've already patched.
+	patchedInstructionIndexes := make(map[int]bool)
+
+	for idx, ins := range p.Instructions {
+		if !ins.IsBuiltinCall() || ins.Constant != int64(asm.FnTracePrintk) {
+			continue // Not a call to bpf_trace_printk, skip
+		}
+		maxLookback := max(0, idx-100) // For safety, don't look back more than 100 instructions
+
+		// We found the call to bpf_trace_printk, now we need to find
+		// the string on the stack and patch it.
+		// For that, find first the value of the second register, which is the second argument
+		// to the call, which is the length of the string.
+		// Example instruction: MovImm dst: r2 imm: 0x0000000000000009 which
+		// sets the length of the formatting to 9
+		var lengthLoadIns *asm.Instruction
+		for i := idx - 1; i >= maxLookback; i-- {
+			candidate := &p.Instructions[i]
+			if candidate.OpCode == movImmOpCode && candidate.Dst == asm.R2 {
```

Comment:
From what I've seen yes, it seems that all calls to log_debug are string literals and the length gets put into that register directly. Just in case I checked what happens with two log_debug calls with the same length:

```
     31: MovImm dst: r1 imm: 0x00000000000a3231
           ; log_debug("12"); // Check with a small word...
         32: StXMemW dst: rfp src: r1 off: -112 imm: 0x0000000000000000
         33: MovReg dst: r1 src: rfp
         34: AddImm dst: r1 imm: 0x-000000000000070
           ; log_debug("12"); // Check with a small word...
         35: MovImm dst: r2 imm: 0x0000000000000004
         36: Call FnTracePrintk
         37: MovImm dst: r1 imm: 0x00000000000a3132
           ; log_debug("21"); // and another of the same length to see what does the compiler with that
         38: StXMemW dst: rfp src: r1 off: -112 imm: 0x0000000000000000
         39: MovReg dst: r1 src: rfp
         40: AddImm dst: r1 imm: 0x-000000000000070
           ; log_debug("21"); // and another of the same length to see what does the compiler with that
         41: MovImm dst: r2 imm: 0x0000000000000004
         42: Call FnTracePrintk
```

I am not sure what would happen if users pass a char* and not a string literal as the argument, but in that case the macro fails to compile so I don't think it would cause us any issue.

Reply:
The problem is that register allocation is not deterministic. Very small changes to the code drastically effects register allocation in the bytecode.
We cannot really predict if the case I mentioned above is impossible. It may not be likely, but register pressure can break the assumption in the code.